Support RenderMan spline parameters

danieldresser-ie opened the pull request:
Building on top of the prior Cortex and Gaffer PRs, we should now be ready to take a look at actually supporting PRMan spline parameters.

The part of this that could most be questioned is that I ended up reworking OSLShader pretty heavily - the prior approach of testing separately for each parameter whether we could build a spline using it ( so that the spline is constructed repeated times for each parameter that makes it up ) ... does not work well for the PRMan convention where the count plug has no suffix ( so every plug could potentially be the count of a spline, and you'd have to search for matching parameters ). Instead, I now identify splines starting from the position plug, finding all related parameters, and then adding those parameters to a set so they won't be processed separately. Overall, I think I'm pretty happy with this approach, and as far as I can tell, it matches the old behaviour anywhere it matters, but there are possible changes in behaviour: for example, if there is contradictory metadata specified on multiple parameters matching the same spline, it now overrides in the order "values, then positions, then interpolation", whereas previously this behaviour was undefined. This definitely seems like an improvement, but it is a change.

johnhaddon commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -894,11 +897,16 @@ SplineDefinitionInterpolation basisFromString( const std::string &basis )
 {
-	typename PlugType::ValueType defaultValue;
+	std::string name;
+	std::variant< std::monostate, SplineDefinitionff, SplineDefinitionfColor3f > splineDefault;
```
Suggest splineDefault -> defaultValue.

johnhaddon commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -1163,12 +1246,76 @@ Plug *loadShaderParameter( const OSLQuery &query, const OSLQuery::Parameter *par
+
+		SplineParameters splineP = splineParametersFromPositions( shaderName, query, positionsParameter, prefix, parametersAlreadyProcessed );
+
+		// This is a very weird way of saying "The variant isn't empty", but it seems to be the modern C++ convention )
```
Perhaps this might be clearer if `splineDefault` only had the two _valid_ types, and `splineParametersFromPositions()` returned `std::optional<SplineParameters>`? It's more obvious that a value isn't useable if you can't even _get_ the value, rather than getting it and checking that its fields seem to have what you want.

johnhaddon replied:
The "very weird" comment is still there.

johnhaddon commented on python/GafferRenderManTest/RenderManShaderTest.py:
```diff
@@ -289,5 +289,45 @@ def testDot( self ) :
+
+		self.assertEqual( shader['parameters']['targetHue'].getName(), referencePlug.getName() )
+		self.assertEqual( shader['parameters']['targetHue'].getFlags(), referencePlug.getFlags() )
+		self.assertEqual( shader['parameters']['targetHue'].direction(), referencePlug.direction() )
```
Why make a reference plug, instead of just asserting direction and flags directly?

johnhaddon commented on python/GafferRenderManTest/RenderManShaderTest.py:
```diff
@@ -289,5 +289,45 @@ def testDot( self ) :
+					( 0.833, imath.Color3f( 1, 0.25, 1 ) ),
+					( 1, imath.Color3f( 1, 0.25, 0.25 ) ),
+				),
+				Gaffer.SplineDefinitionInterpolation( 1 )
```
The named value would be clearer than the numeric constant.

johnhaddon commented on src/GafferRenderMan/RenderManShader.cpp:
```diff
@@ -372,24 +360,278 @@ Gaffer::Plug *loadParameter( const boost::property_tree::ptree &parameter, Plug
+				IECore::Msg::Warning, "RenderManShader::loadShader",
+				fmt::format( "Error while parsing parameter \"{}\" : {}", param.first, e.what() )
+			);
+			continue;
```
This breaks the natural ordering of the parameters as they're specified in the `.args` file. See below, where the mode plug is now in the middle rather than the top :

<img width="501" height="150" alt="image" src="https://github.com/user-attachments/assets/c10617a9-79c0-4525-a057-06b7dfff4173" />
 
As well as getting the right ordering for the basic parameters, we need to make sure that the spline parameters are interleaved with them correctly.

johnhaddon reviewed (commented):
Thanks Daniel! Comments inline as usual - mostly nitpicky stuff, with the exception of the RenderManShader parameter ordering. Oh, Changes.md needs updating to mention spline support in RenderMan.

And a reminder : PRs with Arnold or RenderMan changes in them need to be made from a branch on `GafferHQ/gaffer`, not a personal fork. We can only run CI with third-party licenses from the main repo. I don't think we can change the source branch of a PR now and I don't want to lose the review comments by opening a new one, so let's just continue here (I've run the tests locally for now).

Cheers...
John

danieldresser-ie commented on python/GafferRenderManTest/RenderManShaderTest.py:
```diff
@@ -289,5 +289,45 @@ def testDot( self ) :
+			)
+		)
+
+		self.assertEqual( shader['parameters']['targetHue'].getName(), referencePlug.getName() )
```
Addressed in [da07d7c](https://github.com/GafferHQ/gaffer/pull/6650/commits/da07d7cd0c76a7e08144901e72fc9a7aa6189328)

danieldresser-ie commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -1422,36 +1585,72 @@ static IECore::ConstCompoundDataPtr metadataGetter( const std::string &key, size
 	CompoundDataPtr parameterMetadata = new CompoundData;
 	metadata->writable()["parameter"] = parameterMetadata;
+
+	std::set<std::string> parametersAlreadyProcessed;
```
Addressed in https://github.com/GafferHQ/gaffer/pull/6650/commits/4a56296bb0d5becb3292b95b6e9a57299b25488b

danieldresser-ie commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -894,11 +897,16 @@ SplineDefinitionInterpolation basisFromString( const std::string &basis )
 
-template <typename PlugType>
-Plug *loadSplineParameters( const OSLQuery::Parameter *positionsParameter, const OSLQuery::Parameter *valuesParameter, const OSLQuery::Parameter *basisParameter, const InternedString &name, Gaffer::Plug *parent )
+struct SplineParameters
```
I've gone with SplinePlugArguments in https://github.com/GafferHQ/gaffer/pull/6650/commits/4a56296bb0d5becb3292b95b6e9a57299b25488b

danieldresser-ie commented:
Addressed feedback. Added Changelog entries. I think it could be ready for a squash and merge once we figure out what we're doing about the TODO on replacePlug ( this simple option is just to remove the call to replacePlug ).

I hadn't really thought explicitly before about what our policy should be for sets/maps when performance doesn't seem important enough to profile ... I suppose it's reasonable to standardize on unordered* for better asymptotic performance, even when the average case is likely to be worse - in the average case, this isn't going to be a performance bottleneck anyway.

johnhaddon reviewed (commented):
> Addressed feedback. Added Changelog entries. I think it could be ready for a squash and merge once we figure out what we're doing about the TODO on replacePlug ( this simple option is just to remove the call to replacePlug ).

Thanks! For the TODO, I propose we just revert to the previous code. Not because it's better, but because its not entirely clear and we have no need to change it now (we're already changing a bit more than we're completely comfortable with).

I also added an inline comment about the map of parameters in RenderManShader - I think we should be able to avoid that - might be nice to do two iterations over the parameters instead of three.

> I hadn't really thought explicitly before about what our policy should be for sets/maps when performance doesn't seem important enough to profile ... I suppose it's reasonable to standardize on unordered* for better asymptotic performance, even when the average case is likely to be worse - in the average case, this isn't going to be a performance bottleneck anyway.

That sounds like a reasonable rule, and I'm happy with where the PR is at container-wise. But my assumption was that unordered wins for the average case too - I feel like I've profiled it for something to prove this, but can't recall the details now. I also assumed unordered was better for data locality. Am I wrong?
